monkey/monkey_island/cc/server_utils/encryption/data_store_encryptor.py

```diff
@@ -20,61 +23,43 @@
 
-    def init_key(self, secret: str):
-        if os.path.exists(self.key_file_path):
-            self._load_existing_key(secret)
+    def __init__(self, key_file_path: str, secret: str):
+        if os.path.exists(key_file_path):
+            self._key_based_encryptor = DataStoreEncryptor._load_existing_key(key_file_path, secret)
         else:
-            self._create_new_key(secret)
+            self._key_based_encryptor = DataStoreEncryptor._create_new_key(key_file_path, secret)
 
-    def _load_existing_key(self, secret: str):
-        with open(self.key_file_path, "rb") as f:
+    @staticmethod
+    def _load_existing_key(key_file_path: str, secret: str):
+        with open(key_file_path, "rb") as f:
             encrypted_key = f.read()
-        cipher_key = (
-            PasswordBasedByteEncryptor(secret).decrypt(io.BytesIO(encrypted_key)).getvalue()
-        )
-        self._key_base_encryptor = KeyBasedEncryptor(cipher_key)
-
-    def _create_new_key(self, secret: str):
-        cipher_key = Random.new().read(self._BLOCK_SIZE)
-        encrypted_key = (
-            PasswordBasedByteEncryptor(secret).encrypt(io.BytesIO(cipher_key)).getvalue()
-        )
-        with open_new_securely_permissioned_file(self.key_file_path, "wb") as f:
+        cipher_key = PasswordBasedBytesEncryptor(secret).decrypt(encrypted_key)
+        return KeyBasedEncryptor(cipher_key)
+
+    @staticmethod
+    def _create_new_key(key_file_path: str, secret: str):
+        cipher_key = Random.new().read(DataStoreEncryptor._BLOCK_SIZE)
+        encrypted_key = PasswordBasedBytesEncryptor(secret).encrypt(cipher_key)
+        with open_new_securely_permissioned_file(key_file_path, "wb") as f:
             f.write(encrypted_key)
-        self._key_base_encryptor = KeyBasedEncryptor(cipher_key)
-
-    def is_key_setup(self) -> bool:
-        return self._key_base_encryptor is not None
+        return KeyBasedEncryptor(cipher_key)
 
     def enc(self, message: str):
-        return self._key_base_encryptor.encrypt(message)
+        return self._key_based_encryptor.encrypt(message)
 
     def dec(self, enc_message: str):
-        return self._key_base_encryptor.decrypt(enc_message)
+        return self._key_based_encryptor.decrypt(enc_message)
 
 
-def initialize_datastore_encryptor(key_file_dir: str):
+def initialize_datastore_encryptor(username: str, password: str):
     global _encryptor
 
-    _encryptor = DataStoreEncryptor(key_file_dir)
-
-
-class EncryptorNotInitializedError(Exception):
-    pass
-
-
-def setup_datastore_key(secret: str):
-    if _encryptor is None:
-        raise EncryptorNotInitializedError
-    else:
-        if not _encryptor.is_key_setup():
-            _encryptor.init_key(secret)
+    factory = get_encryptor_factory()
+    if not factory:
+        raise FactoryNotInitializedError
+    secret = get_secret_from_credentials(username, password)
+    _encryptor = DataStoreEncryptor(factory.key_file_path, secret)
```

Comment:
Factories are responsible for constructing objects. We should have `factory.set_key_file_path()` and `factory.set_secret()` and then call `factory.construct_data_store_encryptor()` or similar.

Reply:
Yes, this is not a factory, it's a class that handles initialization parameters for data store encryptor. I'm not sure what would be a better name, maybe `DataStoreEncryptorInitializer`? I don't see any benefits in making this an actual factory, since the output object will not be different depending on key file path and secret. What is more, the usage of such factory is less intuitive. The registration resource would have to `factory.set_secret()` then `built_encryptor = factory.construct_data_store_encryptor()` and finally `set_encryptor(built_encryptor)`. Now it's only one call - `initialize_datastore_encryptor(username, password)`.

Reply:
I think we can simplify this way down. I'm going to dig into it a bit.

Reply:
I'll try to refactor and see how it looks